### ISSUE TITLE
TraceMismatchError when using trace decorator

### ISSUE THREAD

**Issue by JackSnowWolf** [MegEngine/MegEngine#163]
Hi!

I faced this error when I followed instruction on the MegEngine Page to train a simple CIFAR-10 model.

Here is the error:

![lALPD3lGtpC6AxXNAdvNArU_693_475](https://user-images.githubusercontent.com/35722743/113292675-6906e800-9327-11eb-99a0-d394adc9ed2b.png)
![lALPD3W5N-vwRDvNAW7NAeE_481_366 png_620x10000q90g](https://user-images.githubusercontent.com/35722743/113292688-6b694200-9327-11eb-8a4a-6a1ae9db8b4c.jpg)

It seems the JIT part doesn't work when I use `@trace(symbolic=True)` or `@trace(symbolic=False)` during Training or testing. I could run the code when I remove this decorator. It seems the JIT doesn't work. I was using Python 3.8 and Ubuntu eviroment 18.04. Where Can I find procedure to install JIT for MegEngine correctly?

Thanks!

**Comment by yzchen** [comment]
trace accepts tensor as input not numpy array, so you can do following before calling to traced function:

```
data = mge.tensor(data)
label = mge.tensor(label)
```

then it will work as expected.

**Comment by JackSnowWolf** [comment]
> trace accepts tensor as input not numpy array, so you can do following before calling to traced function:
> 
> ```
> data = mge.tensor(data)
> label = mge.tensor(label)
> ```
> 
> then it will work as expected.


I don't think this problem is due to not correct type. When I use `trace` decorator in prediction, it should use `@trace(symbolic=True, capture_as_const=True)`

The input pred and label have already been converted to `tensor`.
After I changed the decorator input, the problem is fixed.

But, I strongly suggest you to update the instruction on the official website.

Thank you!

**Comment by JackSnowWolf** [comment]
@megvii-mge

**Comment by yzchen** [comment]
> 
> 
> > trace accepts tensor as input not numpy array, so you can do following before calling to traced function:
> > ```
> > data = mge.tensor(data)
> > label = mge.tensor(label)
> > ```
> > 
> > 
> > then it will work as expected.
> 
> I don't think this problem is due to not correct type. When I use `trace` decorator in prediction, it should use `@trace(symbolic=True, capture_as_const=True)`
> 
> The input pred and label have already been converted to `tensor`.
> After I changed the decorator input, the problem is fixed.
> 
> But, I strongly suggest you to update the instruction on the official website.
> 
> Thank you!

Happy to see you have solved the problem!
However, I want to mention something here: it's not necessary to use `symbolic=True` and `capture_as_const=True` specifically. In general `symbolic(True)` and `symbolic(False)` should give you the same results, they are just different ways for tracing which lead to the same endpoint. `capture_as_const` is required when you want to dump your model, if not, then it's unnecessary.

BUT, We do need to update our online tutorials, that's a very good suggestion!!! @MegChai

**Comment by MegChai** [comment]
@yzchen  Thanks for the mention! Good idea. Actually, I had updated the doc website just a few hours ago before this issue was created. Some content has been reorganized.

According to @JackSnowWolf 's feedback, 
it seems that the current document is not clear enough for the explanation of the parameters that the `trace` accepts. I am planning to add some explanation about `capture_as_const` on [trace](https://megengine.org.cn/doc/stable/zh/user-guide/model-development/trace.html) page and give a link to [dump](https://megengine.org.cn/doc/stable/zh/user-guide/model-development/dump.html) page. 

--------------------
There is no need to set the parameter `capture_as_const=True` each time while we using `jit.trace` .

> The input pred and label have already been converted to tensor.

I'm not sure about this behavior in given codes so it might need to show more details.

**Comment by JackSnowWolf** [comment]
Thank you for your response!

Actually, I tried several times. 
- The first time, I only use `@trace` during training and predicting. It's okay to use `trace` during training, but there is an error when predicting, which is `TraceMismatchError`.
- The second time, I remove that decorator and run the script successfully.
- After that, I use `@trace(symbolic=True, capture_as_const=True)` during predicting and run the script successfully.
- Then, @yzchen and @MegChai talked about the input type issue. [#163 (comment)](https://github.com/MegEngine/MegEngine/issues/163#issuecomment-812275712) I tried to input a `tensor` `pred` and an `nd.array` `label`. But this time I got a different error, which is about invalid operation between type tensor and nd.array.

@MegChai I'm not very sure about that whether `capture_as_const` is required during predicting, but the exception is reported inside the `eval_func`. Maybe I dump the model after this step, but I guess this operation is unknown or unpredicting in this stage. I guess such kind of error should be reported in the dump operation. If you can locate the error, that that's great.

**Comment by MegChai** [comment]
@JackSnowWolf  I guess I have got your point.

- The first time, you explicitly converted the input training data and label to `Tensor` type in the training code but forgot to do the same work during predicting. This may be due to some old-fashioned official tutorials, our tutorial is keeping updated recently. Now we clearly know the tutorial should remind the user to convert the type.
- The second time, you remove that decorator. In dynamic graph mode, the input `ndarray` will be implicitly converted to `Tensor`.  (But not for all operators, so we still suggest do explicitly conversion) Everything works.
- `capture_as_const=True` will capture global vars or closures as const value including `ndarray` vars, so it will make thing seems right. But it was not the recommended way.
- The last situation: " I tried to input a tensor pred and an nd.array label. "

Note that the `label` should also be converted to `Tensor`:

```
data = mge.tensor(data)
label = mge.tensor(label)  # Notice this line
```

Please have a try again~ 

-----------

**Could you please give us the complete code here? So we can reproduce the error.** It's the most efficient way to find the actual reason. By the way, there are many examples in MegEngine/[Models](Models) that might be helpful.


I think this bad user experience is caused by our incomplete docs, welcome to give us more suggestions.

**Comment by JackSnowWolf** [comment]
I guess I've already got your point. The misunderstanding is mainly due to unclear error message. I guess It should report invalid error type instead of `TraceMismatchError`. Maybe it's a good idea to also add implicit type convert for trace option.

Thank you for your answer and your updating the official page.
